### PR TITLE
[FLINK-12485][python] Add completeness test for Table and TableEnviro…

### PR DESCRIPTION
**Pull request by sunjincheng121**
## What is the purpose of the change

After FLINK-12407 Python Table API will align with current Java Table API. So, In the following Table API development, it's better to add a new user interface to the Java Table API will also need to be added to the Python Table API. So In this Jira, we want to add the test_completeness to check the user interface of Python Table API aligns with Java Table API. 

I'm not sure if we want to force contributors to contribute to the Python Table API while developing the Java Table API. But at least we should have the ability to automatically check if Python is aligned with Java functionality.

## Brief change log
  -  Add `PythonAPICompletenessTestCase` which is base class for Python API completeness tests.
  - Add `TableAPICompletenessTests` which test whether the Python `Table` is consistent with Java `org.apache.flink.table.api.Table`.
  - Add `EnvironmentAPICompletenessTests` which test whether the Python `TableEnvironment` is consistent with Java `org.apache.flink.table.api.TableEnvironment`.

## Verifying this change
This change is only adding the test case.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature?  (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: flink-python/pyflink/testing/test_case_utils.py [hunk 1]
```diff
@@ -90,3 +92,82 @@ def setUp(self):
...
+    @classmethod
+    @abstractmethod
+    def java_class(cls):
+        """
+        Return the Java class that needs to be compared. such as `org.apache.flink.table.api.Table`.
+        """
+        pass
+
+    @classmethod
+    def exclude_methods(cls):
```
Comment (dianfu) [inline flink-python/pyflink/testing/test_case_utils.py]: exclude_methods -> excluded_methods

File: flink-python/pyflink/testing/test_case_utils.py [hunk 2]
```diff
@@ -90,3 +92,82 @@ def setUp(self):
...
+        return {cls.snake_to_camel(cls.java_method_name(method_name))
+                for method_name in dir(python_class) if not method_name.startswith('_')}
+
+    @staticmethod
+    def snake_to_camel(method_name):
+        output = ''.join(x.capitalize() or '_' for x in method_name.split('_'))
+        return output[0].lower() + output[1:]
+
+    @staticmethod
+    def get_java_class_method(java_class):
```
Comment (dianfu) [inline flink-python/pyflink/testing/test_case_utils.py]: get_java_class_method -> get_java_class_methods

File: flink-python/pyflink/testing/test_case_utils.py [hunk 3]
```diff
@@ -90,3 +92,82 @@ def setUp(self):
...
+        if len(missing_methods) > 0:
+            print(missing_methods)
+            print('The Exception should be raised after FLINK-12407 is merged.')
+            # raise Exception('Methods: %s in Java class %s have not been added in Python class %s.'
+            #                % (missing_methods, java_class, python_class))
+
+    @classmethod
+    def java_method_name(cls, python_method_name):
+        """
+        This method should be overWrite when the method name of the Python API cannot be
```
Comment (dianfu) [inline flink-python/pyflink/testing/test_case_utils.py]: overWrite  -> overwritten

File: flink-python/pyflink/testing/test_case_utils.py [hunk 4]
```diff
@@ -90,3 +92,82 @@ def setUp(self):
...
+        consistent with the Java API method name. we need to get the corresponding
```
Comment (dianfu) [inline flink-python/pyflink/testing/test_case_utils.py]: What about removing this line?
`we need to get the corresponding Java API method name based on the Python method name.`

File: flink-python/pyflink/testing/test_case_utils.py [hunk 5]
```diff
@@ -90,3 +92,82 @@ def setUp(self):
...
+        Java API method name based on the Python method name. e.g.: 'as' is python
+        keyword, so we use 'alias' in Python API corresponding 'as' in Java API.
+
+        :param python_method_name: Method name of Python API.
+        :return: Method name corresponding to Java API.
```
Comment (dianfu) [inline flink-python/pyflink/testing/test_case_utils.py]: The corresponding method name of Java API.

File: flink-python/pyflink/testing/test_case_utils.py [hunk 6]
```diff
@@ -90,3 +92,82 @@ def setUp(self):
...
+    def exclude_methods(cls):
+        """
+        Exclude method names that do not need to be checked. When adding excluded methods to the lists
+        you should give a good reason in a comment.
+        :return:
+        """
+        return {}
+
+    def test_completeness(self):
+        if self.python_class() is not None:
```
Comment (dianfu) [inline flink-python/pyflink/testing/test_case_utils.py]: if self.python_class() is not None and self.java_class() is not None

File: flink-python/pyflink/testing/test_case_utils.py [hunk 7]
```diff
@@ -90,3 +92,82 @@ def setUp(self):
...
+    @classmethod
+    def check_methods(cls):
+        java_methods = PythonAPICompletenessTestCase.get_java_class_method(cls.java_class())
+        python_methods = cls.get_python_class_methods(cls.python_class())
+        missing_methods = java_methods - python_methods - cls.exclude_methods()
+        if len(missing_methods) > 0:
+            print(missing_methods)
+            print('The Exception should be raised after FLINK-12407 is merged.')
+            # raise Exception('Methods: %s in Java class %s have not been added in Python class %s.'
+            #                % (missing_methods, java_class, python_class))
```
Comment (dianfu) [inline flink-python/pyflink/testing/test_case_utils.py]: cls.java_class(), cls.python_class()

**Comment by sunjincheng121** [comment]
Thanks for the quick review! @dianfu 

I have update the PR according your comments!

Best,
Jincheng

File: flink-python/pyflink/table/tests/test_environment_completeness.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
+# ###############################################################################
```
Comment (dianfu) [inline flink-python/pyflink/table/tests/test_environment_completeness.py]: Removes the empty space here?

File: flink-python/pyflink/table/tests/test_table_completeness.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+if __name__ == '__main__':
+    import unittest
+
+    try:
+        import xmlrunner
+        testRunner = xmlrunner.XMLTestRunner(output='target/test-reports')
+    except ImportError:
+        testRunner = None
+    unittest.main(testRunner=testRunner, verbosity=2)
```
Comment (dianfu) [inline flink-python/pyflink/table/tests/test_table_completeness.py]: Add one empty line at the end of file

Comment (sunjincheng121) [reply flink-python/pyflink/table/tests/test_table_completeness.py]: Already there :)

**Comment by sunjincheng121** [comment]
Merging...
@flinkbot approve all

**Comment by sunjincheng121** [comment]
Will be merged after CI turn green.
@flinkbot approve all

**Comment by yanghua** [comment]
@sunjincheng121 There is a very high-frequency test failure in table module may cause the Travis to fail. I have reported it, the Jira ID is FLINK-12512. I also found it caused [your PR to fail](https://travis-ci.org/apache/flink/builds/532617753). Can you try to fix it? thanks.

**Comment by sunjincheng121** [comment]
> @sunjincheng121 There is a very high-frequency test failure in table module may cause the Travis to fail. I have reported it, the Jira ID is FLINK-12512. I also found it caused [your PR to fail](https://travis-ci.org/apache/flink/builds/532617753). Can you try to fix it? thanks.

We know the fail reason, Python API add the code style check: you can see the error log  `./pyflink/testing/test_case_utils.py:164:101: E501 line too long (102 > 100 characters) `
And it already fixed in my last commit in this PR. :)  but thanks for your kind mention that.

**Comment by yanghua** [comment]
@sunjincheng121 I mean `TableSourceTest ` test error. It also caused your PR's failure, you can see here : https://travis-ci.org/apache/flink/jobs/532617757

**Comment by sunjincheng121** [comment]
Do you think this problem is caused by this PR? Otherwise, we can discuss it in the JIRA you created, what do you think?

**Comment by yanghua** [comment]
@sunjincheng121 No, I am sorry to comment these unrelated messages in this PR. I just saw it also affected your PR and you guys focus on Table modules may fix it as soon as possible. Because it has broken my PRs many time recently.

**Comment by sunjincheng121** [comment]
So, can we merge this PR first, and then talk the test issue on the JIRA page? @yanghua

**Comment by yanghua** [comment]
@sunjincheng121 Absolutely, it's OK.

**Pull request closed without merging**